node/core/dispute-coordinator/src/lib.rs

```diff
@@ -349,6 +351,103 @@ where
 	}
 }
 
+// Restores the subsystem's state before proceeding with the main event loop. Primarily, this
+// repopulates the rolling session window the relevant session information to handle incoming
+// import statement requests.
+//
+// This method also retransmits a DisputeParticiationMessage::Participate for any non-concluded
+// disputes for which the subsystem doesn't have a local statement, ensuring it eventually makes an
+// arbitration on the disupte.
+async fn resume<B, Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	backend: &mut B,
+) -> Result<(), Error>
+where
+	Context: overseer::SubsystemContext<Message = DisputeCoordinatorMessage>,
+	Context: SubsystemContext<Message = DisputeCoordinatorMessage>,
+	B: Backend,
+{
+
+	// Query for the set of active leaves to reconstruct our rolling session window.
+	let (tx, rx) = oneshot::channel();
+	ctx.send_message(ChainSelectionMessage::Leaves(tx)).await;
+	let leaves = rx.await?;
+
+	// Process the leaves as we would during normal operation. Any writes are committed
+	// immediately so that they are made available to the primary event loop.
+	let mut overlay_db = OverlayedBackend::new(backend);
+	handle_new_activations(ctx, &mut overlay_db, state, leaves).await.unwrap();
+	let write_ops = overlay_db.into_write_ops();
+	backend.write(write_ops)?;
+
+	let recent_disputes = match backend.load_recent_disputes() {
+		Ok(Some(disputes)) => disputes,
+		Ok(None) => return Ok(()),
+		Err(e) => {
+			tracing::error!(target: LOG_TARGET, "failed initial load of recent disputes: {:?}", e);
+			return Err(e.into());
+		},
+	};
+
+	// Filter out disputes that have already concluded.
+	let active_disputes = recent_disputes.into_iter()
+		.filter(|((_, _), status)| status.concluded_at().is_none())
+		.collect::<RecentDisputes>();
+
+	for ((session, ref candidate_hash), _) in active_disputes.into_iter() {
+		let votes: CandidateVotes = match backend.load_candidate_votes(session, candidate_hash) {
+			Ok(Some(votes)) => votes.into(),
+			Ok(None) => continue,
+			Err(e) => {
+				tracing::error!(target: LOG_TARGET, "failed initial load of candidate votes: {:?}", e);
+				continue;
+			},
+		};
+
+		let validators = match state.rolling_session_window.session_info(session) {
+			None => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					session,
+					"Missing info for session which has an active dispute",
+				);
+
+				continue;
+			}
+			Some(info) => info.validators.clone(),
+		};
+
+		let n_validators = validators.len();
+		let voted_indices: HashSet<_> = votes.voted_indices().into_iter().collect();
+		for (index, validator) in validators.iter().enumerate() {
+			let index = ValidatorIndex(index as _);
+			if voted_indices.contains(&index) { continue }
+
+			// Skip non-local validators.
+			if state.keystore.key_pair::<ValidatorPair>(validator).ok().flatten().is_none() {
```

Comment:
this is probably copied from somewhere, but we should use `has_keys` instead of `key_pair.is_none()`